### ISSUE TITLE
Executor not finding service scripts

### ISSUE THREAD

**Issue by dgeorgievski** [keptn-sandbox/generic-executor-service#6]
I am using the latest generic-executor-service v0.3 with keptn v0.7.3.

Keptn project is created as 
```
keptn create project wpng \
--shipyard=./shipyard.yml \
--git-remote-url=https://github.com/**** \
--git-token=****** \
--git-user=*****
```
The content of shipyard.yaml
```
stages:
  - name: "dev"
    approval_strategy:
      pass: automatic
      warning: manual
    deployment_strategy: "direct"

```

Create service 
```
keptn create service event-engine --project=wpng
```

Add resources
```
# deployment script
keptn add-resource \
--project=wpng \
--stage=dev \
--service=event-engine \
--resource=configuration.change.py \
--resourceUri=generic-executor/configuration.change.py

# deployment finished web hook
keptn add-resource \
--project=wpng \
--stage=dev \
--service=event-engine \
--resource=deployment.finished.http \
--resourceUri=generic-executor/deployment.finished.http
```

Send config change event
```
keptn send event new-artifact \
--project=wpng \
--service=event-engine \
--image=myimage \
--tag=mytag
```

The `generic-executor-service` logs show
```
generic-executor-service-5bb7d76cfd-h6s56 generic-executor-service {"timestamp":"2020-12-08T04:15:51.746158157Z","logLevel":"DEBUG","message":"Keptn Resource not found or empty: wpng//event-engine/generic-executor/configuration.change.py - Resource not found"}
```
It looks like the stage information (dev) is not used by `generic-executor-service`. I went quickly through the code consuming the incoming event 
https://github.com/keptn/go-utils/blob/v0.7.1/pkg/lib/keptn.go#L62-L66

I am not sure why the Stage info is missing at the moment. I'd expect the shipyard service would supply this information.

The config-service is correctly saving the scripts under correct location and branch (stage) name.
`/data/config/wpng/event-engine/generic-executor`

**Comment by grabnerandi** [comment]
Hi. I just looked at your issue and think I know what the problem is.
The issue is that when you send a "new artifact" event via the Keptn CLI you only specify service and project but no stage. The Keptn CLI therefore sends a `sh.keptn.event.configuration.change` with an empty stage - this is why the generic executor also doesnt know which stage to look into. The stage itself is added by Keptn at a later stage when it analysis the shipyard. I am sure you can observe the same in your Keptns Bridge when you look at the individual events that are coming in. The first "configuration change" will not have a stage - but the following events will

Long story short: the behavior of the generic executor service is correct. The problem is that the stage is not added by the keptn CLI when sending an event. I therefore think this is either a CLI issue - or - the Keptn API that receives that event initially should "fill in" the stage by taking it from the shipyard

**Comment by dgeorgievski** [comment]
True the `sh.keptn.event.configuration.change` didn't set the Stage. 
```
{
  "contenttype": "application/json",
  "data": {
    "canary": {
      "action": "set",
      "value": 100
    },
    "eventContext": null,
    "labels": null,
    "project": "wpng",
    "service": "event-engine",
    "stage": "",
    "valuesCanary": {
      "image": "myimage:mytag"
    }
  },
  "id": "227d11e6-6b82-4fce-bbb6-2742fc58686f",
  "source": "https://github.com/keptn/keptn/cli#configuration-change",
  "specversion": "0.2",
  "time": "2020-12-07T23:28:42.212Z",
  "type": "sh.keptn.event.configuration.change",
  "shkeptncontext": "a86f5ad3-9cbc-4acf-a6ba-6b151be56f00"
}
```

I am still learning about keptn. Based on the documentation I expected that keptn would provided the stages based on the shipyard.yaml stages definition.

Would it be possible to submit config event directly to /api/configuration-service/v1/event API? 
https://keptn.sh/docs/0.7.x/reference/api/

**Comment by johannes-b** [comment]
Hi @dgeorgievski, 

as @grabnerandi already pointed out, the first configuration.change event does not contain the stage information because Keptn assumes that it will be applied for the first stage in the shipyard file (a smart-default assumption). 

However, you can define your own `configuration.change` event **containing** the stage, as you did above: 
```
{
  "contenttype": "application/json",
  "data": {
    "canary": {
      "action": "set",
      "value": 100
    },
    "eventContext": null,
    "labels": null,
    "project": "wpng",
    "service": "event-engine",
    "stage": "development",
    "valuesCanary": {
      "image": "myimage:mytag"
    }
  },
  "id": "227d11e6-6b82-4fce-bbb6-2742fc58686f",
  "source": "https://github.com/keptn/keptn/cli#configuration-change",
  "specversion": "0.2",
  "time": "2020-12-07T23:28:42.212Z",
  "type": "sh.keptn.event.configuration.change",
  "shkeptncontext": "a86f5ad3-9cbc-4acf-a6ba-6b151be56f00"
}
```
When you store this payload in, e.g., configuration-change.json - then you can use the generic CLI command `keptn send event -f configuration-change.json` to send the event to Keptn. (Or you use the Keptn API at the endpoint `/event`).

**Comment by dgeorgievski** [comment]
I missed the `-f` flag for keptn send event.   I just tested this command and it worked.

Couple of questions about the event id  and shkeptncontext values. From what I can tell these are UUID generated values. 

- Each `send event`call should generate new UUID values before making the request, right?
- Is the value of "source" validated? I didn't have a chance to check this in the code. In other words can I add something like "curl"?

One concern is that there is no correlation between the shipyard stages and this service, but I assume this should be addressed in the future.

It would be also great if `keptn send event`  accepts stage as a flag. Some of the developers will not be excited about triggering events with json files.

**Comment by johannes-b** [comment]
Great, that this approach helps you with your use case. 

* Regarding the two properties `id` and `shkeptncontext`: When they are not provided, Keptn will create them automatically. Hence, you don't need to generate UUIDs and can leave them out. 

* The property `source` is not validated and works just as an indicator that tells Keptn from where this event is coming from.

**Comment by dgeorgievski** [comment]
We can close this issue
